pkg/server/exec/exec.go

```diff
@@ -0,0 +1,106 @@
+package exec
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/luizalabs/teresa/pkg/server/app"
+	"github.com/luizalabs/teresa/pkg/server/auth"
+	"github.com/luizalabs/teresa/pkg/server/database"
+	"github.com/luizalabs/teresa/pkg/server/spec"
+	"github.com/luizalabs/teresa/pkg/server/storage"
+)
+
+type Operations interface {
+	Command(user *database.User, appName string, command ...string) (io.ReadCloser, <-chan error)
+	CommandBySpec(podSpec *spec.Pod) (io.ReadCloser, <-chan error)
+}
+
+type K8sOperations interface {
+	DeployAnnotation(namespace, deployName, annotation string) (string, error)
+	PodRun(podSpec *spec.Pod) (io.ReadCloser, <-chan int, error)
+	IsNotFound(err error) bool
+}
+
+type Defaults struct {
+	RunnerImage  string
+	LimitsCPU    string
+	LimitsMemory string
+}
+
+type ExecOperations struct {
+	appOps   app.Operations
+	fs       storage.Storage
+	k8s      K8sOperations
+	defaults *Defaults
+}
+
+func (ops *ExecOperations) Command(user *database.User, appName string, command ...string) (io.ReadCloser, <-chan error) {
+	errChan := make(chan error, 1)
+	a, err := ops.appOps.Get(appName)
+	if err != nil {
+		errChan <- err
+		return nil, errChan
+	}
+	if !ops.appOps.HasPermission(user, appName) {
+		errChan <- auth.ErrPermissionDenied
+		return nil, errChan
+	}
+
+	currentSlug, err := ops.k8s.DeployAnnotation(a.Name, a.Name, spec.SlugAnnotation)
+	if err != nil {
+		if ops.k8s.IsNotFound(err) {
+			errChan <- ErrDeployNotFound
+		} else {
+			errChan <- err
+		}
+		return nil, errChan
+	}
+
+	podSpec := spec.NewRunner(
+		fmt.Sprintf("exec-command-%s", appName),
+		currentSlug,
+		ops.defaults.RunnerImage,
+		a,
+		ops.fs,
+		&spec.ContainerLimits{
+			CPU:    ops.defaults.LimitsCPU,
+			Memory: ops.defaults.LimitsMemory,
+		},
+		command...,
+	)
+
+	return ops.CommandBySpec(podSpec)
+}
+
+func (ops *ExecOperations) CommandBySpec(podSpec *spec.Pod) (io.ReadCloser, <-chan error) {
+	errChan := make(chan error)
+	r, w := io.Pipe()
```

Comment:
Shouldn't it be `make(chan error, 1)`. Otherwise `errChan <- err` will block, right?

Reply:
Yes, but I'm writing in this channel inside a goroutine, I think there's no problem in this case, what do you think about it?

Reply:
No problem, go ahead.